### DISCUSSION TITLE
"getStaticProps" for API endpoints

### DISCUSSION THREAD

**Discussion by Schniz** [vercel/next.js#12101] [Ideas]
# Feature request

Hey! first, thanks for this wonderful project. 👏 
Regarding the feature, I'm not quite sure if this is already supported or not, I couldn't find it on the docs.

## Is your feature request related to a problem?

Imagine building a simple search engine for your blog, website, or whatever. You can use solr, ElasticSearch or Algolia, but you can also build an index using Lunr or another library. This can be built on build time instead of runtime, and have a JSON endpoint up and working.

Obviously, this can be done by having another build process, but having it integrated to Next.js allows ease of development, like static React-based pages.

Another use case for it could be persisting GraphQL queries. Running a build-time function that generates all the available queries, instead of having a separate process or doing that on runtime

## Describe the solution you'd like

I'm not quite sure regarding the API and whether it should be something other than an API. What do you think about the following API, which follows the pages API?

```js
export default (req, res, staticProps) => {
  res.send(`${staticProps.greeting}, world!`);
}

export const getStaticProps = async () => {
 return { greeting: 'hello' }
}
```

## Describe alternatives you've considered

* Allowing pages to return something other than React components: I don't think it's pretty. Pages are cool and it's very cool that only `/api` files are different
* pre-build actions: won't be watched, degraded developer experience

**Comment by mrclay** [comment]
This would be another approach to get to the same end: **Allow a JSON file as API route to be exported** #17961
